libvast/vast/format/bgpdumpbinary.hpp

```diff
@@ -0,0 +1,1455 @@
+#include "vast/concept/parseable/string.hpp"
+#include "vast/concept/parseable/numeric/byte.hpp"
+#include "vast/concept/parseable/vast/address.hpp"
+#include "vast/concept/parseable/vast/time.hpp"
+#include "vast/concept/parseable/vast/subnet.hpp"
+#include "vast/concept/parseable/vast/data.hpp"
+#include "vast/concept/printable/vast/address.hpp"
+#include "vast/concept/printable/vast/subnet.hpp"
+#include "vast/concept/printable/vast/data.hpp"
+#include "vast/concept/printable/string.hpp"
+#include "vast/concept/printable/to_string.hpp"
+#include "vast/concept/printable/std/chrono.hpp"
+
+#include "vast/format/reader.hpp"
+
+namespace vast {
+namespace format {
+namespace bgpdumpbinary {
+
+struct form {
+  timestamp ts;
+  count bgp_type;
+  count type;
+  count subtype;
+  count interface_index;
+  count addr_family;
+  count old_state;
+  count new_state;
+  count bgp_length;
+  count length;
+  count pasnr;
+  count med;
+  count local_pref;
+  count version;
+  count my_autonomous_system;
+  count hold_time;
+  count bgp_id;
+  count error_code;
+  count error_subcode;
+  std::string msg_type;
+  std::string origin;
+  std::string as_path_orded;
+  std::string community;
+  std::string atomic_aggregate;
+  address peer_ip_v4;
+  address peer_ip_v6;
+  address nexthop_v4;
+  address nexthop_v6;
+  std::vector<data> as_path;
+  std::vector<subnet> prefix_v4;
+  std::vector<subnet> prefix_v6;
+  std::tuple<count, address> aggregator;
+
+  /* extended fields */
+  uint16_t wd_rts_len;
+  bool paket_error;
+  std::vector<form> rib_entries;
+};
+
+/// A parser that reading bgp messages from the MRT files.
+struct bgpdumpbinary_parser : parser<bgpdumpbinary_parser> {
+  using attribute = event;
+
+  bgpdumpbinary_parser();
+
+  template <class Iterator>
+  bool parse(Iterator &f, Iterator const &l, unused_type) const {
+    if(f >= l)
+      return false;
+    else
+      return true;
+  }
+
+  /*-----------------MRT Header-----------------*/
+  template <class Iterator, class Attribute>
+  bool parse_mrt_header(Iterator &f, Iterator const &l, Attribute &a) const {
+    using namespace parsers;
+    uint16_t t16 = 0;
+    uint32_t t32 = 0;
+
+
+    if(f + 12 <= l) {
+      // MRT - Timestamp
+      b32be.parse(f, f + 4, t32);
+      a.ts = vast::timestamp{std::chrono::seconds{t32}};
+      // MRT - Type
+      b16be.parse(f, f + 2, t16);
+      a.type = count{t16};
+      // MRT - Subtype
+      t16 = 0;
+      b16be.parse(f, f + 2, t16);
+      a.subtype = count{t16};
+      // MRT - Length
+      t32 = 0;
+      b32be.parse(f, f + 4, t32);
+      a.length = count{t32};
+      if(a.type == 17){
+        VAST_DEBUG(this, "MRT MICROSECOND TIMESTAMP not supported");
+        // 32bit microsecond timestamp included in length field calculation
+        f += 4; 
+        a.length -= 4;
+      }
+      return true;
+    } else {
+      VAST_DEBUG(this, "MRT HEADER size exceeded");
+      f = l;
+      return false;
+    }
+  }

```

Comment:
We should aim for a declarative form of the parser, as opposed to imperatively going through the input. See [the parser for IP addresses](https://github.com/vast-io/vast/blob/master/libvast/vast/concept/parseable/vast/address.hpp) as an example. Otherwise it's nearly impossible to verify the manual bound checks and iterator/pointer arithmetic.

If you need missing parser types to operate on binary data, I'm happy to implement them for you. (For example, a parser for packed or varbyte-encoded integers.)